diff --git a/drivers/net/e1000/e1000.h b/drivers/net/e1000/e1000.h
index a881dd0..b18edd5 100644
--- a/drivers/net/e1000/e1000.h
+++ b/drivers/net/e1000/e1000.h
@@ -319,6 +319,9 @@ enum e1000_state_t {
 	__E1000_TESTING,
 	__E1000_RESETTING,
 	__E1000_DOWN
+#if defined(CONFIG_NETMAP) || defined(CONFIG_NETMAP_MODULE)
+        , __E1000_NETMAP
+#endif
 };
 
 #undef pr_fmt
diff --git a/drivers/net/e1000/e1000_main.c b/drivers/net/e1000/e1000_main.c
index 4035a42..6267385 100644
--- a/drivers/net/e1000/e1000_main.c
+++ b/drivers/net/e1000/e1000_main.c
@@ -30,6 +30,10 @@
 #include <net/ip6_checksum.h>
 #include <linux/io.h>
 
+#if defined(CONFIG_NETMAP) || defined(CONFIG_NETMAP_MODULE)
+#  define HAVE_NETMAP
+#endif
+
 /* Intel Media SOC GbE MDIO physical base address */
 static unsigned long ce4100_gbe_mdio_base_phy;
 /* Intel Media SOC GbE MDIO virtual base address */
@@ -182,6 +186,11 @@ static void e1000_shutdown(struct pci_dev *pdev);
 static void e1000_netpoll (struct net_device *netdev);
 #endif
 
+#ifdef HAVE_NETMAP
+#include <net/netmap/if_e1000_netmap.h>
+#endif /* HAVE_NETMAP */
+
+
 #define COPYBREAK_DEFAULT 256
 static unsigned int copybreak __read_mostly = COPYBREAK_DEFAULT;
 module_param(copybreak, uint, 0644);
@@ -1175,6 +1184,10 @@ static int __devinit e1000_probe(struct pci_dev *pdev,
 	if (err)
 		goto err_register;
 
+#ifdef HAVE_NETMAP
+        e1000_netmap_attach(adapter);
+#endif /* HAVE_NETMAP */
+        
 	/* print bus type/speed/width info */
 	e_info(probe, "(PCI%s:%dMHz:%d-bit) %pM\n",
 	       ((hw->bus_type == e1000_bus_type_pcix) ? "-X" : ""),
@@ -1817,6 +1830,15 @@ static void e1000_configure_rx(struct e1000_adapter *adapter)
 	struct e1000_hw *hw = &adapter->hw;
 	u32 rdlen, rctl, rxcsum;
 
+#ifdef HAVE_NETMAP
+        if (test_bit(__E1000_NETMAP, &adapter->flags)) {
+		rdlen = adapter->rx_ring[0].count *
+		        sizeof(struct e1000_rx_desc);
+                adapter->clean_rx = e1000_netmap_clean_rx;
+                adapter->alloc_rx_buf = e1000_netmap_alloc_rx_buffers;
+        }
+        else
+#endif
 	if (adapter->netdev->mtu > ETH_DATA_LEN) {
 		rdlen = adapter->rx_ring[0].count *
 		        sizeof(struct e1000_rx_desc);
diff --git a/include/linux/if.h b/include/linux/if.h
index d901a8c..d375669 100644
--- a/include/linux/if.h
+++ b/include/linux/if.h
@@ -52,9 +52,10 @@
 #define IFF_DORMANT	0x20000		/* driver signals dormant	*/
 
 #define IFF_ECHO	0x40000		/* echo sent packets		*/
+#define IFF_NETMAP      0x80000         /* driver supports netmap       */
 
 #define IFF_VOLATILE	(IFF_LOOPBACK|IFF_POINTOPOINT|IFF_BROADCAST|IFF_ECHO|\
-		IFF_MASTER|IFF_SLAVE|IFF_RUNNING|IFF_LOWER_UP|IFF_DORMANT)
+	IFF_MASTER|IFF_SLAVE|IFF_RUNNING|IFF_LOWER_UP|IFF_DORMANT|IFF_NETMAP)
 
 /* Private (from user) interface flags (netdevice->priv_flags). */
 #define IFF_802_1Q_VLAN 0x1             /* 802.1Q VLAN device.          */
diff --git a/include/linux/netdevice.h b/include/linux/netdevice.h
index 3af28fb..21bfa6b 100644
--- a/include/linux/netdevice.h
+++ b/include/linux/netdevice.h
@@ -950,6 +950,10 @@ struct net_device
 	void			*ax25_ptr;	/* AX.25 specific data */
 	struct wireless_dev	*ieee80211_ptr;	/* IEEE 802.11 specific data,
 						   assign before registering */
+#if defined(CONFIG_NETMAP) || defined(CONFIG_NETMAP_MODULE)
+        void                    *netmap_ptr;
+#endif
+
 
 /*
  * Cache line mostly used on receive path (including eth_type_trans())
diff --git a/net/Kconfig b/net/Kconfig
index 041c35e..73d61c3 100644
--- a/net/Kconfig
+++ b/net/Kconfig
@@ -276,4 +276,11 @@ source "net/wimax/Kconfig"
 source "net/rfkill/Kconfig"
 source "net/9p/Kconfig"
 
+config NETMAP
+       tristate "netmap direct user space NIC access"
+       default m
+       help
+         Netmap provides safe and efficient access to network interfaces
+         from user space.  See http://info.iet.unipi.it/~luigi/netmap/
+
 endif   # if NET
diff --git a/net/Makefile b/net/Makefile
index 1542e72..792a0c5 100644
--- a/net/Makefile
+++ b/net/Makefile
@@ -65,3 +65,4 @@ ifeq ($(CONFIG_NET),y)
 obj-$(CONFIG_SYSCTL)		+= sysctl_net.o
 endif
 obj-$(CONFIG_WIMAX)		+= wimax/
+obj-$(CONFIG_NETMAP)		+= netmap/
